nexus-and-alo/nexus-strawman.md

```diff
@@ -0,0 +1,361 @@
+
+Outside of creating a Nexus Service Instance, a general purpose CRUD API is provided that enables users to:
+
+- List instantiated Nexus Service Instances
+- Remove previously instantiated Nexus Service Instances
+- Replace a Nexus Service Instance definition
+- Get info about an existing Nexus Service Instance
+
+### Nexus Task hosting
+
+Now that the Nexus Service is registered with a Nexus Backend, we have to define, register and run a Nexus Task Handler in order to handle the incoming requests to our endpoint.
+
+The Nexus payloads will be passed through the system before eventually being picked up by an appropriate Nexus Task Worker. Just as with Workflow and Activity Tasks, Nexus Tasks are not parsed or deeply introspected before being delivered. This also means that DataConverter functionality is unhindered and can be fully used with Nexus. 
+
+Here is a proposed API of the Nexus Task Handler:
+
+// TODO (Chad): show Nexus Handler without HTTP (Temporal specific).
+
+```go
+package main
+
+import (
+  "context"
+  "encoding/json"
+  "fmt"
+  "net/http"
+
+  "github.com/cretz/nexus-poc/sdk/go/nexus/backend/client"
+  "github.com/cretz/nexus-poc/sdk/go/nexus/backend/worker"
+)
+
+func HandleMyServiceRequest(ctx context.Context, req *worker.CallRequest) *worker.CallResponse {
+  // Validate HTTP
+  if req.HTTP == nil {
+    return worker.ErrInvalidProtocol
+  } else if req.HTTP.RelativePath != "/foo" {
+    // Note the relative path has the service stripped
+    return worker.NewHTTPError(http.StatusNotFound)
+  } else if req.HTTP.Request.Method != "GET" {
+    return worker.NewHTTPError(http.MethodNotAllowed)
+  } else if req.HTTP.Request.Header.Get("Content-Type") != "application/json" {
+    return worker.NewHTTPErrorf(http.StatusBadRequest, "Invalid input")
+  }
+  
+  // Read body
+  reqBody := struct{ Name string `json:"name"` }{}
+  b, err := io.ReadAll(req.HTTP.Request.Body)
+  if closeErr := req.HTTP.Request.Body.Close(); closeErr != nil && err == nil {
+    err = closeErr
+  }
+  if err == nil {
+    err = json.Unmarshal(b, &reqBody)
+  }
+  if err != nil {
+    return worker.NewHTTPErrorf(http.StatusBadRequest, "Invalid input: %w", err)
+  }
+
+  // Respond with foo JSON body
+  b, _ := json.Marshal(map[string]string{"foo": fmt.Sprintf("Hello, %v!", reqBody.Name)})
+  return worker.NewHTTPOKResponse(b, "Content-Type", "application/json")
+}
+```
+
+**What’s possible within a Nexus Task Handler?**
+
+Within a Nexus Task Handler, the rules are similar to that of an Activity. Nexus Task Handlers enables:
+
+- Making external RPC calls or other I/O operations.
+- Using a full-fledged Temporal client for kicking off Workflows or interacting with other Temporal resources.
+- Interacting with the originating Request through a formal Response.
+
+### Nexus Task Worker
+
+In order to make forward progress on incoming Nexus Tasks, a valid Nexus Worker for the Nexus Instance must be registered.
+
+// TODO (Chad): improve if needed
```

Comment:
I think this gets a bit too much into implementation. I don't think it's fair to design what this Go API looks exactly like now. I suppose it'd be very similar to HTTP serving w/ HTTP handlers in Go and different in other languages.

Reply:
Look thats completely ok but I need to show something. Can you maybe just redo all the samples with pseudocode? I really don't care about the specific interfaces, I just want to demonstrate the capabilities.

Reply:
Let's just go with the example that exists below then. No improvement needed for the proposal.